src/dbnode/persist/fs/cross_block_reader_test.go

```diff
@@ -0,0 +1,170 @@
+	dfsReader1.EXPECT().Range().Return(xtime.Range{Start: now})
+
+	later := now.Add(time.Hour)
+	dfsReader2 := NewMockDataFileSetReader(ctrl)
+	dfsReader2.EXPECT().IsOrderedByIndex().Return(true)
+	dfsReader2.EXPECT().Range().Return(xtime.Range{Start: later})
+
+	_, err := NewCrossBlockReader([]DataFileSetReader{dfsReader2, dfsReader1})
+
+	expectedErr := fmt.Errorf("dataFileSetReaders are not ordered by time (%s followed by %s)", later, now)
+
+	assert.Equal(t, expectedErr, err)
+}
+
+func TestCrossBlockReader(t *testing.T) {
+	tests := []struct {
+		name           string
+		blockSeriesIds [][]string
+	}{
+		{"no readers", [][]string{}},
+		{"empty readers", [][]string{{}, {}, {}}},
+		{"one reader, one series", [][]string{{"id1"}}},
+		{"one reader, many series", [][]string{{"id1", "id2", "id3"}}},
+		{"many readers with same series", [][]string{{"id1"}, {"id1"}, {"id1"}}},
+		{"many readers with different series", [][]string{{"id1"}, {"id2"}, {"id3"}}},
+		{"many readers with unordered series", [][]string{{"id3"}, {"id1"}, {"id2"}}},
+		{"complex case", [][]string{{"id2", "id3", "id5"}, {"id1", "id2", "id4"}, {"id1", "id4"}}},
+		{"immediate reader error", [][]string{{"error"}}},
+		{"reader error later", [][]string{{"id1", "id2"}, {"id1", "error"}}},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			testCrossBlockReader(t, tt.blockSeriesIds)
+		})
+	}
+}
+
+func testCrossBlockReader(t *testing.T, blockSeriesIds [][]string) {
+	ctrl := xtest.NewController(t)
+	defer ctrl.Finish()
+
+	now := time.Now().Truncate(time.Hour)
+	var dfsReaders []DataFileSetReader
+	expectedCount := 0
+
+	for blockIndex, ids := range blockSeriesIds {
+		dfsReader := NewMockDataFileSetReader(ctrl)
+		dfsReader.EXPECT().IsOrderedByIndex().Return(true)
+		dfsReader.EXPECT().Range().Return(xtime.Range{Start: now.Add(time.Hour * time.Duration(blockIndex))})
+
+		blockHasError := false
+		for j, id := range ids {
+			tags := ident.NewTags(ident.StringTag("foo", string(j)))
+			data := checkedBytes([]byte{byte(j)})
+			checksum := uint32(blockIndex) // somewhat hacky - using checksum to propagate block index value for assertions
+			if id == "error" {
+				dfsReader.EXPECT().Read().Return(nil, nil, nil, uint32(0), expectedError)
+				blockHasError = true
+			} else {
+				dfsReader.EXPECT().Read().Return(ident.StringID(id), ident.NewTagsIterator(tags), data, checksum, nil)
+			}
+		}
+
+		if !blockHasError {
+			dfsReader.EXPECT().Read().Return(nil, nil, nil, uint32(0), io.EOF).MaxTimes(1)
+		}
+
+		dfsReaders = append(dfsReaders, dfsReader)
+		expectedCount += len(ids)
+	}
+
+	cbReader, err := NewCrossBlockReader(dfsReaders)
+	require.NoError(t, err)
+	defer cbReader.Close()
+
+	actualCount := 0
+	previousId := ""
+	for cbReader.Next() {
+		id, tags, records := cbReader.Current()
+
+		strId := id.String()
+		id.Finalize()
+		assert.True(t, strId > previousId, "series must be read in increasing id order")
```

Comment:
nit: may be more obvious what this is doing if this tested against a list of expected IDs, e.g.
```
{
 name: "many readers with unordered series", 
 blockSeriesIDs: [][]string{{"id3"}, {"id1"}, {"id2"}}},
 expectedIDs: []string{"id1","id2","id3"}
 }
```
Then here append the ids to a slice as you iterate, and compare them after hitting EOF

Reply:
Not sure about this. Would have to add loads of repetitive boilerplate. Please let me know if you still think this is a better idea.

Reply:
`loads of repetitive boilerplate` seems to be the motto of testing in go :p IMO if it's just the ordered ID list that's being checked, it may be worthwhile having it explicitly, just because the current approach may make it difficult to debug the root cause quickly, the current kind of check would definitely work well for a prop test though :+1:

Reply:
👍